Add IPv6 triple colon fix

gfr10598 opened the pull request:
soltesz primary
yachang FYI

1.  Adds a repair for triple colon IP address errors.
2.  Moves util stuff to web100 package
3.  Changes various SS parse errors from abort to continue.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/etl/467)
<!-- Reviewable:end -->

stephen-soltesz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[parser/util.go, line 37 at r1](https://reviewable.io:443/reviews/m-lab/etl/467#-L6xFaqJdAber5Ax01es:-L6xFaqJdAber5Ax01et:bx4nmca) ([raw file](https://github.com/m-lab/etl/blob/6a7dab1ca07f815d2ffb20fc41c90a9c3307ae06/parser/util.go#L37)):*
> ```Go
> )
> 
> // FixBadIPv6 fixes triple colon ::: which is produced by sidestream.
> ```

produced by older versions of the c-web100 library (used by sidestream).

paris-traceroute (via rollins) has avoided this bug due to the fact that paris traceroute doesn't support ipv6 (afaik).

---

*[parser/util.go, line 38 at r1](https://reviewable.io:443/reviews/m-lab/etl/467#-L6xFXRdAlp1PrDIuwP_:-L6xFXRdAlp1PrDIuwPa:b-5g4iut) ([raw file](https://github.com/m-lab/etl/blob/6a7dab1ca07f815d2ffb20fc41c90a9c3307ae06/parser/util.go#L38)):*
> ```Go
> 
> // FixBadIPv6 fixes triple colon ::: which is produced by sidestream.
> func FixBadIPv6(ipStr string) (string, error) {
> ```

Suggest, `NormalizeIPv6`. 

---

*[parser/util.go, line 41 at r1](https://reviewable.io:443/reviews/m-lab/etl/467#-L6xE9URREdIxt3E2HMh:-L6xE9USPq5DE5ip5Dnz:b-cremvb) ([raw file](https://github.com/m-lab/etl/blob/6a7dab1ca07f815d2ffb20fc41c90a9c3307ae06/parser/util.go#L41)):*
> ```Go
> 	split := strings.Split(ipStr, ":::")
> 	switch len(split) {
> 	case 0:
> ```

Can `strings.Split()` ever return an empty list? I believe the zero case is not possible, even if ipStr is empty.

---

*[parser/util.go, line 51 at r1](https://reviewable.io:443/reviews/m-lab/etl/467#-L6xHIRgHtTtp9c5ZRML:-L6xHIRgHtTtp9c5ZRMM:b49tk90) ([raw file](https://github.com/m-lab/etl/blob/6a7dab1ca07f815d2ffb20fc41c90a9c3307ae06/parser/util.go#L51)):*
> ```Go
> 		return split[0] + "::" + split[1], nil
> 	default:
> 		return ipStr, ErrIPv6MultipleTripleColon
> ```

We do not have any sites with `::` in the /64 prefix. If we did, e.g. "2001::34::/64", it could be possible to have two "triple-colons" in the IP. I'm not totally sure. I believe this is the code with the bug -- https://github.com/npad/web100-userland/blob/00b9681d17b6ab2ff694e5560819a290ace95bde/lib/web100.c#L1175 as called by the python wrapper https://github.com/npad/web100-userland/blob/00b9681d17b6ab2ff694e5560819a290ace95bde/python/Web100.py#L254

Could we collapse any pattern matching ":::+" to "::" safely?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/467#-:-L6xK-8U12Lupb-yyyF5:bnfp4nl)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[parser/util.go, line 37 at r1](https://reviewable.io:443/reviews/m-lab/etl/467#-L6xFaqJdAber5Ax01es:-L6xKXyrJ8Y9vQaQrN_Q:b-5e96hi) ([raw file](https://github.com/m-lab/etl/blob/6a7dab1ca07f815d2ffb20fc41c90a9c3307ae06/parser/util.go#L37)):*
> produced by older versions of the c-web100 library (used by sidestream).
Added first comment.

---

*[parser/util.go, line 38 at r1](https://reviewable.io:443/reviews/m-lab/etl/467#-L6xFXRdAlp1PrDIuwP_:-L6xKmb_Q-JhWq6g6waW:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/6a7dab1ca07f815d2ffb20fc41c90a9c3307ae06/parser/util.go#L38)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Suggest, `NormalizeIPv6`. 
</blockquote></details>

Done.

---

*[parser/util.go, line 41 at r1](https://reviewable.io:443/reviews/m-lab/etl/467#-L6xE9URREdIxt3E2HMh:-L6xKvjakNB_nDHW82jZ:bl9mdgm) ([raw file](https://github.com/m-lab/etl/blob/6a7dab1ca07f815d2ffb20fc41c90a9c3307ae06/parser/util.go#L41)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Can `strings.Split()` ever return an empty list? I believe the zero case is not possible, even if ipStr is empty.
</blockquote></details>

Yes, but only "If both s and sep are empty, Split returns an empty slice."  Of course, not the case in this case, so I've adjusted code.

---

*[parser/util.go, line 51 at r1](https://reviewable.io:443/reviews/m-lab/etl/467#-L6xHIRgHtTtp9c5ZRML:-L6xLRFM7ngceFFrpLYb:b-9flzqy) ([raw file](https://github.com/m-lab/etl/blob/6a7dab1ca07f815d2ffb20fc41c90a9c3307ae06/parser/util.go#L51)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

We do not have any sites with `::` in the /64 prefix. If we did, e.g. "2001::34::/64", it could be possible to have two "triple-colons" in the IP. I'm not totally sure. I believe this is the code with the bug -- https://github.com/npad/web100-userland/blob/00b9681d17b6ab2ff694e5560819a290ace95bde/lib/web100.c#L1175 as called by the python wrapper https://github.com/npad/web100-userland/blob/00b9681d17b6ab2ff694e5560819a290ace95bde/python/Web100.py#L254

Could we collapse any pattern matching ":::+" to "::" safely?
</blockquote></details>

Probably, but this seems more conservative, and will give us more info if there are things we don't expect.  So prefer as is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/467)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/467#-:-L6xMnruNiOWPjsgCeZi:bnfp4nl)*
<!-- Sent from Reviewable.io -->
